price_pusher/src/sui/sui.ts

```diff
@@ -0,0 +1,247 @@
+          .price.fields.conf;
+
+      const timestamp =
+        priceInfoObject.data.content.fields.price_info.fields.price_feed.fields
+          .price.fields.timestamp;
+
+      return {
+        price: negative ? "-" + magnitude : magnitude,
+        conf,
+        publishTime: Number(timestamp),
+      };
+    } catch (e) {
+      console.error(`Polling Sui on-chain price for ${priceId} failed. Error:`);
+      console.error(e);
+      return undefined;
+    }
+  }
+}
+
+export class SuiPricePusher implements IPricePusher {
+  private readonly signer: RawSigner;
+  constructor(
+    private priceServiceConnection: PriceServiceConnection,
+    private pythPackageId: string,
+    private pythStateId: string,
+    private wormholePackageId: string,
+    private wormholeStateId: string,
+    private priceFeedToPriceInfoObjectTableId: string,
+    endpoint: string,
+    mnemonic: string
+  ) {
+    this.signer = new RawSigner(
+      Ed25519Keypair.deriveKeypair(mnemonic),
+      new JsonRpcProvider(new Connection({ fullnode: endpoint }))
+    );
+  }
+
+  async updatePriceFeed(
+    priceIds: string[],
+    pubTimesToPush: number[]
+  ): Promise<void> {
+    if (priceIds.length === 0) {
+      return;
+    }
+
+    if (priceIds.length !== pubTimesToPush.length)
+      throw new Error("Invalid arguments");
+
+    const tx = new TransactionBlock();
+
+    const vaas = await this.priceServiceConnection.getLatestVaas(priceIds);
+
+    // Parse our batch price attestation VAA bytes using Wormhole.
+    // Check out the Wormhole cross-chain bridge and generic messaging protocol here:
+    //     https://github.com/wormhole-foundation/wormhole
+    let verified_vaas: any = [];
+    for (const vaa of vaas) {
+      const [verified_vaa] = tx.moveCall({
+        target: `${this.wormholePackageId}::vaa::parse_and_verify`,
+        arguments: [
+          tx.object(this.wormholeStateId),
+          tx.pure([...Buffer.from(vaa, "base64")]),
+          tx.object(SUI_CLOCK_OBJECT_ID),
+        ],
+      });
+      verified_vaas = verified_vaas.concat(verified_vaa);
+    }
+
+    // Create a hot potato vector of price feed updates that will
+    // be used to update price feeds.
+    let [price_updates_hot_potato] = tx.moveCall({
+      target: `${this.pythPackageId}::pyth::create_price_infos_hot_potato`,
+      arguments: [
+        tx.object(this.pythStateId),
+        tx.makeMoveVec({
+          type: `${this.wormholePackageId}::vaa::VAA`,
+          objects: verified_vaas,
+        }),
+        tx.object(SUI_CLOCK_OBJECT_ID),
+      ],
+    });
+
+    // Update each price info object (containing our price feeds of interest)
+    // using the hot potato vector.
+    for (const priceId of priceIds) {
+      let priceInfoObjectId;
+      try {
+        priceInfoObjectId = await priceIdToPriceInfoObjectId(
+          this.signer.provider,
+          this.pythPackageId,
+          this.priceFeedToPriceInfoObjectTableId,
+          priceId
+        );
```

Comment:
i'm pretty sure this mapping is static, so you could precompute it once and then use the cached version. That would reduce the number of API calls that this code makes.